aws-cni: L7 Ingress proxy support for Security Group attached Pods in chaining mode

andjhop opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!


---
This is a proposal for one possible fix for #29922, I'm aware that this particular approach may not be deemed acceptable. In short, Cilium layer 7 policies are incomptabile with security group attached pods (SGP pods) in chaining mode with the AWS VPC CNI; the use of `iif` rules in the routing policy configuration for SGP pods means local traffic from Envoy is not routed to its intended destination. Below, a pod at `10.10.10.33/24` makes a request to another pod at `10.10.10.22/24` with a layer 7 ingress policy applied:

```mermaid
flowchart BT
    trunk(
        🔗 eth1#nbsp;#0040;trunk#0041;
        10.10.10.11/24) 

    vlan[🔗 vlan.eth.1@eth1]
    rt[[
         Routing#nbsp;Policy
         fwmark#nbsp;0x200/0xF00#nbsp;lookup#nbsp;2004
         iif#nbsp;vlan.eth.1#nbsp;lookup#nbsp;101
         iif#nbsp;vlan3bee14957f5#nbsp;lookup#nbsp;101
         from#nbsp;all#nbsp;lookup#nbsp;main
         <br>Route#nbsp;Table#nbsp;2004
         local#nbsp;default#nbsp;dev#nbsp;lo
         <br>Route#nbsp;Table#nbsp;101
         default#nbsp;via#nbsp;10.10.10.1#nbsp;dev#nbsp;vlan.eth.1
         10.10.10.1#nbsp;dev#nbsp;vlan.eth.1
         10.10.10.22#nbsp;dev#nbsp;vlan3bee14957f5
        <br>Route#nbsp;Table#nbsp;Main
        default#nbsp;via#nbsp;10.10.10.1#nbsp;dev#nbsp;eth0
        10.10.10.0/24#nbsp;dev#nbsp;eth0#nbsp;src#nbsp;10.10.10.1]]

    tc((🐝))
    podeth
    proxy((Envoy Proxy))
    eni[🔗 vlan3bee14957f5]
    subgraph pod [pod]
        podeth(
            🔗 eth0
            10.10.10.22/24)
    end

    trunk --- vlan
    vlan -->|1#nbsp;src:#nbsp;10.10.10.10.33#nbsp;dst:#nbsp;10.10.10.22| rt
    rt -->|2| tc --- eni --- podeth
    tc -->|3#nbsp;fwmark#nbsp;0x0200/0x0F00| rt
    rt -->|4| proxy
    proxy -->|5#nbsp;fwmark#nbsp;0x0A00/0x0F00| rt
    rt -->|6#nbsp;src:#nbsp;10.10.10.11#nbsp;dst:#nbsp;10.10.10.22| trunk
```
When security groups are not involved, the network setup is quite generic and requires no special consideration. Conversely, when a security group is attached to a pod, In addition to its veth pair (`vlan3bee14957f5`/`eth0`), the AWS CNI will create a corresponding VLAN device (`vlan.eth.«VLANID»`) with a unique VLAN ID (1 in this example). Traffic to or from `10.10.10.22/24` is always expected to ingress in to and egress out of the node on `vlan.eth.1` even when the peer is another pod on the same node. A special route table is created for the SGP pod with the table ID `100 + «VLANID»`(101) along with the following routing policy rules:
```
iif vlan.eth.1 lookup 101
iif vlan3bee14957f5 lookup 101
```
Traffic to the destination pod `10.10.10.22/24` would ordinarily have arrived on `vlan.eth.1`(1) resulting in a lookup of route table 101 and delivery out of `vlan3bee14957f5`(2). Since traffic sourced from Envoy(5) is local, `iif vlan.eth.1 lookup 101` doesn't apply, and traffic instead takes the default route back out of the node(6). The solution proposed here is to install additional routing policy rules so ingress proxy traffic marked `0x0A00` is routed using the SGP pod route table (101 in the example). This would result in the example above in the following rules being installed:
```
from all to 10.10.10.22 fwmark 0x0A00/0x0F00 lookup 101
from 10.10.10.22 to all fwmark 0x0A00 lookup 101
```
Now traffic originating from Envoy(5) to the destination pod `10.10.10.22/24` and replies to traffic originating from the remote peer `10.10.10.33/24` and terminating at Envoy(4) are routed using table 101, mirroring the setup without Envoy in the loop. This behaviour will be configured with a new `--cni-proxy-routing` command line option.

### Consideration 1: Reply traffic from destination pod to ingress proxy
For handling reply traffic from the destination pod `10.10.10.22/24` on a connection sourced from Envoy, we need to be able to distinguish it from traffic sourced from the pod itself. As such, I've added a new field to `ct_entry`: `from_ingress_proxy`. Reply traffic from the destination pod back to Envoy has `from_ingress_proxy` set to true and is marked with the "to proxy" magic `0x0200` for local routing. 

### Consideration 2: Reverse path filtering
This setup requires special iptables rules in order to preserve reverse path filtering. Firstly, replies from the destination pod arriving on `vlan3bee14957f5` back to Envoy would usually fail reverse path filtering as the symmetric routing rule in the opposite direction specifies a firewall mark(`0x0A00`) and so is not considered for source interface validation. Traffic sourced from the remote peer destined to Envoy arriving on `vlan.eth.1` fails similarly.

gandro commented on api/v1/openapi.yaml:
```diff
@@ -2731,6 +2731,9 @@ definitions:
       ipLocalReservedPorts:
         description: Comma-separated list of IP ports should be reserved in the workload network namespace
         type: string
+      CNIProxyRouting:
```
Have you investigated if this could be part of the CNI config file instead? https://github.com/cilium/cilium/blob/f6606c6ccdb80c187243ec8359ddfb037aa75969/plugins/cilium-cni/types/types.go#L22

My understanding is that this file is available even in chaining mode. It is written here: https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/daemon/cmd/cni/config.go#L274

(Though I might be missing something here, I'm not intimately familiar with the CNI config lifecycle in chaining mode)

andjhop replied:
I've made it part of the CNI config file now, it didn't even occur to me to do it this way.

gandro reviewed (commented):
Thanks for the PR and apologies for the late review! I lack context on the proxy rules, so I cannot judge correctness of the installed IPTable rules. I mainly reviewed the CNI lifecycle parts.

While I'm not a fan of assuming too much of the behavior of AWS VPC CNI (e.g. interface nameing etc), if we have no better solution for this, then it is what it is.

Two additional non-code related comments:
 - Please note that CI is failing because the branch is out of date.
 - Your diagrams don't render correctly on GitHub. Could you provide rendered versions of them?

andjhop committed:
bpf: Track ingress proxy connections

Add a new field to the conntrack entry struct for tracking connections
originating from an ingress proxy.

Signed-off-by: Andrew Hopkins <andrew.hopkins@starlingbank.com>

andjhop committed:
aws-cni: Install ingress proxy rules

Install special ingress proxy rules for security group attached pods

Signed-off-by: Andrew Hopkins <andrew.hopkins@starlingbank.com>

andjhop committed:
iptables: AWS CNI rpfilter rules

Install special reverse path filter rules for security group attached
pods

Signed-off-by: Andrew Hopkins <andrew.hopkins@starlingbank.com>

andjhop committed:
aws-cni: make special network setup for SGP Pods optional

Signed-off-by: Andrew Hopkins <andrew.hopkins@starlingbank.com>

andjhop committed:
aws-cni: Move proxy routing configuration to the CNI NetConf

Signed-off-by: Andrew Hopkins <andrew.hopkins@starlingbank.com>

andjhop commented on plugins/cilium-cni/chaining/awscni/network.go:
```diff
@@ -0,0 +1,91 @@
+// disableIfaceRPFilter disables the `rp_filter` network setting for the
+// network interface iface
+func disableIfaceRPFilter(iface string) error {
+	return os.WriteFile(
```
With IPv4, we need to disable reverse path filtering by the kernel and handle it instead with iptables. Traffic sourced from the remote peer `10.10.10.33/24` incoming on `vlan.eth.1`(1) would be dropped otherwise. This is because the routing rule that would direct traffic in the opposite direction out over the same interface, specifies the firewall mark(`0x0A00`) and so isn't considered. Replies from the destination pod `10.10.10.22/24` on connections source from Envoy(2) would be dropped similarly.

andjhop commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
+		"-m", "mark", "--mark", fmt.Sprintf("%#08x", linux_defaults.MagicMarkIsProxy),
+		"-j", "CONNMARK", "--set-mark", ctMarkIsProxy}); err != nil {
+		return err
+	}
```
Mark connections incoming on `vlan.eth.1`(1), sourced from the remote peer, destined to Envoy as a "proxy connection".

andjhop commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
+		"-m", "rpfilter", "--loose", "--accept-local",
+		"-j", "ACCEPT"}); err != nil {
+		return err
+	}
```
Incoming traffic from the remote peer on `vlan.eth.1`(1) and from the destination pod on `vlan3bee14957f5`(2) can be accepted with no further checks if it passes an rpfilter check.

andjhop commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
+		"-i", "vlan.eth.+",
+		"-j", "DROP"}); err != nil {
+		return err
+	}
```
Incoming traffic from the remote peer on `vlan.eth.1`(1) that fails an rpfilter check should be accepted if this is a "proxy connection".

andjhop commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
+		"-m", "mark", "!", "--mark", fmt.Sprintf("%#08x", linux_defaults.MagicMarkIsToProxy),
+		"-m", "rpfilter", "--loose", "--invert",
+		"-j", "DROP"})
+}
```
Incoming traffic from the destination pod on `vlan3bee14957f5`(2) that fails an rpfilter check should be accepted if it's reply traffic to the proxy. All other traffic should be dropped.

andjhop reviewed (commented):
```mermaid
flowchart LR
    vlan[🔗 vlan.eth.1@eth1]
    proxy((Envoy Proxy))
    eni[🔗 vlan3bee14957f5]
    subgraph pod [pod]
        podeth(
            🔗 eth0
            10.10.10.22/24)
    end

    vlan <-->|1 src:&nbsp;10.10.10.10.33 dst:&nbsp;10.10.10.22| proxy
    proxy <-->|2 src:&nbsp;10.10.10.10.11 dst:&nbsp;10.10.10.22| eni <--> podeth
 ```

gandro commented on plugins/cilium-cni/chaining/awscni/network.go:
```diff
@@ -0,0 +1,91 @@
+	return os.WriteFile(
+		fmt.Sprintf("/proc/sys/net/ipv4/conf/%s/rp_filter", iface),
+		[]byte("0"),
+		0644)
```
nit: Could we use `sysctl.Write` here? Here's an example of how this can be done from the CNI plugin: https://github.com/cilium/cilium/blob/82a8f409acc6bee8ce71f444ac1d892d8775383b/plugins/cilium-cni/cmd/cmd.go#L424

dylandreimerink replied:
I second this, its preferred reuse the same API whenever we can. Since the CNI isn't in the agent you can't use the reconciling sysctl logic, but we have a fallback available: https://github.com/cilium/cilium/blob/d49bec4b6e047d55b01efeffdad18a338907936e/pkg/datapath/linux/sysctl/sysctl.go#L181-L196

We even have an existing copy of this very logic on tap: https://github.com/cilium/cilium/blob/d49bec4b6e047d55b01efeffdad18a338907936e/pkg/datapath/connector/add.go#L48

dylandreimerink replied:
Oh, I only now noticed you fixed this in a later commit (I am reviewing one by one). Can you rebase this old commit instead so we keep a clean git history?

gandro commented on plugins/cilium-cni/chaining/awscni/result.go:
```diff
@@ -0,0 +1,51 @@
+// isSGPPodAttachment returns true if this is an attachment for a
+// security group attached Pod following strict enforcement
+func isSGPPodAttachment(res *cniTypesVer.Result) bool {
+	return strings.HasPrefix(getHostIface(res), awsCNIIfacePrefixSGPPod)
```
This just checks if the first interface in the result has to correct prefix.

But is it guaranteed that a) the first interface exists and b) that it is the host index interface? Seems a bit fragile if the order changes or if the interface list is empty for some reason. 

In particular if it's empty, the CNI plugin will crash with an array out of bounds.

The CNI result does contain a `sandbox` field that if empty indicates that it's a host-network level interface. Would it make sense to iterate over the returned interfaces and validate both the name and the "sandbox" field? And basically return the index of the matching interface here, rather than just a boolean?

Ref: https://www.cni.dev/docs/spec/#success

gandro reviewed (commented):
Thanks a lot for the additional explanations! I think high-level this seems reasonable, at least as far as I understand it. 

If left some additional nits that I discovered in a second passthrough. Other than that, I think this seems to look good to me.

andjhop committed:
aws-cni: use existing sysctl functionality

Signed-off-by: andjhop <andrew.hopkins@starlingbank.com>

andjhop committed:
aws-cni: soften assumptions about AWS CNI Result format

Signed-off-by: andjhop <andrew.hopkins@starlingbank.com>

gandro reviewed (approved):
Thank you!

asauber reviewed (commented):
Hi, thanks for the PR!

Have you considered the implications of SGPP "standard" mode vs "strict" mode with this change? https://aws.github.io/aws-eks-best-practices/networking/sgpp/#enforcing-mode-use-standard-mode-in-the-following-situations.

One thing that I notice about standard mode is that the device prefix of `vlan.` is not used.

asauber reviewed (approved):
Approving for SIG CLI. The flag implementation and documentation seems fine.

Can you expand on why you say the following in your description?

> I'm aware that this particular approach may not be deemed acceptable

Is it because this adds another set of iptables rules in this scenario?

squeed commented on daemon/cmd/cni/cell.go:
```diff
@@ -73,6 +80,7 @@ func (cfg Config) Flags(flags *pflag.FlagSet) {
 	flags.String(option.CNIChainingTarget, defaultConfig.CNIChainingTarget, "CNI network name into which to insert the Cilium chained configuration. Use '*' to select any network.")
 	flags.Bool(option.CNIExclusive, defaultConfig.CNIExclusive, "Whether to remove other CNI configurations")
 	flags.Bool(option.CNIExternalRouting, defaultConfig.CNIExternalRouting, "Whether the chained CNI plugin handles routing on the node")
+	flags.Bool(option.CNIProxyRouting, defaultConfig.CNIProxyRouting, "Install L7 Proxy routes where necessary even when cni-external-routing is set")
```
I wonder if there's more we can say about this here (or somewhere else reasonable). Imagine a "standard" chained case; we don't need this flag. Why? (I understand why from the PR, but it's not clear from this help entry).

squeed commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1568,6 +1568,12 @@ func (m *Manager) installRules(state desiredState) error {
 		}
 	}
 
+	if m.cniConfigManager.GetChainingMode() == "aws-cni" && m.cniConfigManager.ProxyRoutingEnabled() {
```
Please add a comment explaining why this is being done :-)

squeed commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
 		"-j", "CONNMARK", "--restore-mark", "--nfmask", nfmask, "--ctmask", ctmask})
 }
+
+func (m *Manager) addAWSCNIRPFilterRules() error {
```
This function needs a serious docblock :-)

squeed commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1794,3 +1800,65 @@ func (m *Manager) addCiliumENIRules() error {
+	// Track connections which originate from a remote peer and terminate at an
+	// ingress proxy. This traffic will ingress into the node on vlan.eth.+,
+	// replies from the proxy will egress out of the node on vlan.eth.+, this
+	// would still fail rpfilter however as the reverse FIB lookup can't
```
Without understanding the whole context, should we also be manipulating the rp_filter sysctl?

Personally, I'm not so sure; as a chained plugin we don't necessarily "own" the interface sysctls. However, if we need to change a sysctl to make it work, then we should do that :-).

andjhop replied:
Yea I agree, I really didn't want to have to touch these settings.

squeed commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -870,6 +870,9 @@ data:
 {{- if (not (kindIs "invalid" .Values.cni.externalRouting)) }}
   cni-external-routing: {{ .Values.cni.externalRouting | quote }}
 {{- end}}
+{{- if (not (kindIs "invalid" .Values.cni.proxyRouting)) }}
```
This seems to be missing from `values.yaml` -- can you add it, with documentation?

jrajahalme commented on bpf/bpf_lxc.c:
```diff
@@ -768,6 +768,11 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	if (ct_state->from_ingress_proxy) {
+		ctx->mark = MARK_MAGIC_TO_PROXY;
+	} else
+#endif /* ENABLE_REPLY_TO_PROXY_MARK */
```
There is a code block in this same function higher up where this should be taken care of:
```
	case CT_REPLY:
		/* Skip policy enforcement for return traffic. */
		/* Check if this is return traffic to an ingress proxy. */
		if (ct_state->proxy_redirect) {
...
```
It looks like `ct_state->proxy_redirect` is only set for the connection coming into the ingress proxy, and as we never use the same 5-tuple for the forwarded connections from the ingress proxy, this code block actually never matches the return traffic to an ingress proxy as intended. It would seem to me that instead of adding this new conditionally compiled block here we should amend the code block above to read:
```
	case CT_REPLY:
		/* Skip policy enforcement for return traffic. */
		/* Check if this is return traffic to an ingress proxy. */
		if (ct_state->proxy_redirect || ct_state->from_ingress_proxy) {
...
```
The called proxy redirect function `ctx_redirect_to_proxy6(ctx, tuple, 0, false)` already sets `ctx->mark = MARK_MAGIC_TO_PROXY`. This also means that we do not need the new preprocessor symbol `ENABLE_REPLY_TO_PROXY_MARK`

andjhop replied:
Ah, this makes more sense, thank you.

jrajahalme commented on bpf/bpf_lxc.c:
```diff
@@ -1326,6 +1331,11 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
+	if (ct_state->from_ingress_proxy) {
+		ctx->mark = MARK_MAGIC_TO_PROXY;
+	} else
+#endif /* ENABLE_REPLY_TO_PROXY_MARK */
```
Same here, instead of adding conditionally compiled block here, maybe modify the existing block above to also check for the new flag:
```
	case CT_REPLY:
		/* Skip policy enforcement for return traffic. */
		/* Check if this is return traffic to an ingress proxy. */
		if (ct_state->proxy_redirect || ct_state->from_ingress_proxy) {
```

jrajahalme reviewed (changes requested):
Looks like we could check for the new CT flag in an existing code block that currently fails to match the ingress proxy return traffic.

andjhop committed:
Merge remote-tracking branch 'origin/main' into ah/l7-proxy-support-for-aws-cni-sgp-pods

Signed-off-by: andjhop <andrew.hopkins@starlingbank.com>
